src/puppetlabs/puppetdb/query_eng/engine.clj

```diff
@@ -1741,6 +1828,70 @@
 
    :else (throw (IllegalArgumentException. (format "Your initial query must be of the form: [\"from\",<entity>,(<optional-query>)]. Check your query and try again.")))))
 
+(pls/defn-validated ^:private fix-in-expr-multi-comparison
+  "Returns [column projection] after adjusting the type of one of them
+  to match the other if that one is of type :multi and the other
+  isn't."
+  [column :- column-schema
+   projection :- projection-schema]
+  ;; For now we have to assume it's fv.*, etc.
+  (let [[proj-name proj-info] projection
+        multi-col? (= :multi (:type column))
+        multi-proj? (= :multi (:type proj-info))]
+    (cond
+      (and multi-col? multi-proj?)
+      (do
+        (assert (= (:field column) :fv.value))
+        (assert (= (:field proj-info) :fv.value))
+        [column projection])
+      multi-col?
+      (do
+        (assert (= (:field column) :fv.value))
+        [(merge column (fv-variant (:type proj-info)))
+         projection])
+      multi-proj?
+      (do
+        (assert (= (:field proj-info) :fv.value))
+        [column
+         [proj-name (merge proj-info (fv-variant (:type column)))]])
+      :else
+      [column projection])))
+
+(defn- fix-in-expr-multi-comparisons
+  [node]
+  (let [columns (:column node)
+        projected-fields (get-in node [:subquery :projected-fields])]
+    (assert (= (count columns) (count projected-fields)))
+    (loop [cols columns
+           fields projected-fields
+           fixed-cols []
+           fixed-fields []]
+      (if-not (seq cols)
```

Comment:
Sure, if it's confusing -- I suppose to some extent it's a habit; I sometimes tend to put the smaller and/or base case first, and the more complex case, or a recursion, last.